tests/python_package_test/test_basic.py

```diff
@@ -36,9 +74,13 @@ def test(self):
             if i % 10 == 0:
                 print(bst.eval_train(), bst.eval_valid())
 
+        (lower_bound, upper_bound) = find_upper_and_lower_bounds(bst)
+
         self.assertEqual(bst.current_iteration(), 20)
         self.assertEqual(bst.num_trees(), 20)
         self.assertEqual(bst.num_model_per_iteration(), 1)
+        self.assertAlmostEqual(bst.upper_bound(), upper_bound, places=5)
+        self.assertAlmostEqual(bst.lower_bound(), lower_bound, places=5)
```

Comment:
Is it possible to hardcode expected values now?

Reply:
I added it along now! Thanks

Reply:
I tried to add it also, but for lower bound I still see some precision loss even with 4 digits, with 3 digits it works. But since the test with the parsed json tree works, I think it is not anything comming from this feature implementation

Reply:
Hmmm, I'm getting the following numbers:

```
Upper
3.3182142872462883 Linux
3.3182142872462883 macOS
3.3182142872462883 Windows

Lower
-2.9040190126976606 Linux
-2.9040190126976606 macOS
-2.9040190126976606 Windows
```

Please remove all `places=` arguments from asserts and we'll see with what exact numbers it fails.

Reply:
Done, let's see what happens

Reply:
For the lower bound the error appears in the 3rd decimal for the CI tests, but I am  not sure what platform or conditions it fails under

Reply:
The failing tests are GPU ones. Please add `"gpu_use_dp": True` here. Everything should be OK then with any precision.
https://github.com/microsoft/LightGBM/blob/1e5049a1a6bdde4e37701a237b025bfe060e4119/tests/python_package_test/test_basic.py#L22-L30

Reply:
Done, thank you very much for your help @StrikerRUS

Reply:
Thanks a lot for all your hard debugging work! I think we can now remove `find_upper_and_lower_bounds` helper and leave only comparison with hardcoded values in the test.

Reply:
Done! Thanks